### PR TITLE
Error diagnostics for the S3000

### PR DESCRIPTION
**Pull request by ghost**
This expands the S3000 driver to allow it to read the 7-segment error codes directly, without the need for a human to look at the screen.  The check disables continuous data transmission temporarily, reads in the error code, and then waits for continuous data transmission to automatically resume.  This check is triggered by a rostopic published from another node.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -26,7 +26,7 @@ add_library(SerialDevice
   external/SerialDevice/src/SerialDevice.cc
 )
 
-add_executable(s3000_laser src/s3000_laser.cc src/sicks3000.cc)
+add_executable(s3000_laser src/s3000_node.cc src/sicks3000.cc)
```
Comment (ghost) [inline CMakeLists.txt:29]: Hopefully this clears up some of the confusion between these two files.

File: include/sicks3000.h [hunk 2]
```diff
@@ -60,14 +60,17 @@ class SickS3000
     //! Close the port
     int Close();
 
+    //! Check the operating data block of the S3000 for system and error status
+    int CheckOperatingDataBlock(int retry_count);
+
     //! Read and process data
-    int ReadLaser( sensor_msgs::LaserScan& scan_msg, bool& bValidData ); // public periodic function
+    int ReadLaser(sensor_msgs::LaserScan& scan_msg, bool& bValidData); // public periodic function
```
Comment (ghost) [inline include/sicks3000.h:70]: I'm not sure of the styling here, but for now I've at least made the files consistent.  Will need to roslint everything once all the review changes are done.

File: src/sicks3000.cc [hunk 3]
```diff
@@ -310,3 +310,330 @@ unsigned short SickS3000::CreateCRC(uint8_t *Data, ssize_t length)
...
+  SEVEN_SEGMENT_CODE(SEG_u,SEG_9), // *see above (u. - 9.)
+  SEVEN_SEGMENT_CODE(SEG_y,SEG__1), //Internal error in the S3000 (y. - _1.)
+  SEVEN_SEGMENT_CODE(SEG_y,SEG_2), //Internal I/O module error(y. - 2.)
+  SEVEN_SEGMENT_CODE(SEG_y,SEG_3), //I/O module/sensor head device combination invalid (y. - 3.)
+};
+
+const char* s3000_error_causes[NUM_S3000_ERROR_CODES] = {
+  
+  //Indication of protective field and contour infringements in the dual field mode
+  "Object in protective field (1_.)",
```
Comment (mikepurvis) [inline src/sicks3000.cc]: This seems incredibly error-prone, separating the strings from the codes. What was the reason not to use an array of structs?

File: src/sicks3000.cc [hunk 4]
```diff
@@ -310,3 +310,330 @@ unsigned short SickS3000::CreateCRC(uint8_t *Data, ssize_t length)
...
+#define SEG_G   0xBD
+#define SEG_H   0xF6
+#define SEG_L   0xB8
+#define SEG_n   0x54
+#define SEG_o   0xDC
+#define SEG_P   0xF3
+#define SEG_u   0x9C
+#define SEG_y   0xEE
+
+#define SEVEN_SEGMENT_CODE(first, second) (first + (second << 8))
```
Comment (mikepurvis) [inline src/sicks3000.cc]: `#define` seems unnecessary; couldn't all of this have been static consts?

File: src/sicks3000.cc [hunk 5]
```diff
@@ -310,3 +310,330 @@ unsigned short SickS3000::CreateCRC(uint8_t *Data, ssize_t length)
...
+
+#define MAX_READ_ATTEMPTS 10
+int SickS3000::CheckOperatingDataBlock(int retry_count)
+{
+  if (serial!=NULL)
+  {
+    char interrupt_byte[1] = {0x41};
+    char get_token[20] = {0x00,0x00,0x41,0x44,0x19,0x00,0x00,0x05,0xFF,0x07,0x19,0x00,0x00,0x05,0xFF,0x07,0x07,0x0F,0x9F,0xD0};
+    char request_block_11[10] = {0x00,0x00,0x45,0x44,0x0B,0x00,0x00,0x56,0xFF,0x07};
+    char release_token[20] = {0x00,0x00,0x41,0x44,0x19,0x00,0x00,0x05,0xFF,0x07,0x19,0x00,0x00,0x05,0xFF,0x07,0x00,0x00,0xE7,0xB8};
```
Comment (mikepurvis) [inline src/sicks3000.cc]: What are these magic sequences and how did they come about?

File: src/sicks3000.cc [hunk 6]
```diff
@@ -310,3 +310,330 @@ unsigned short SickS3000::CreateCRC(uint8_t *Data, ssize_t length)
...
+      {
+        //Release the system token
+        if(serial->WritePort(release_token, &written_bytes, 20) < 0) {return -1;}
+        if(!serial->BlockOnRead(100)) {return -1;}
+        ros::Duration(0.1).sleep();
+        CheckOperatingDataBlock(++retry_count);
+      }
+      else
+      {
+        ROS_INFO("Unable to read operating data block!");
```
Comment (mikepurvis) [inline src/sicks3000.cc]: Shouldn't this be a warning? ~~Should it retry?~~ I see that there's retries already here, so in that case, the message should be "unable to xxxx after n tries."

File: src/sicks3000.cc [hunk 7]
```diff
@@ -310,3 +310,330 @@ unsigned short SickS3000::CreateCRC(uint8_t *Data, ssize_t length)
...
+      ROS_INFO("device_state %02X display_data %08X led_code %04X", device_state, display_data, (display_data & 0x0000FFFF));
+      
+      for(int i = 0; i < NUM_S3000_ERROR_CODES; i++)
+      {
+        if((display_data & 0x0000FFFF) == display_code[i])
+        {
+          ROS_INFO("%s", s3000_error_causes[i]);
+          ROS_INFO("%s", s3000_error_resolution[i]);
+          break;
+        }
```
Comment (mikepurvis) [inline src/sicks3000.cc]: Case where code is unrecognized?

**Comment by mikepurvis** [comment]
I'm not keen on seeing rosconsole as the primary interface to this functionality— it should be coming up through actual diagnostics. There should be fields for the last need code, how long ago it was read, and the related cause/resolution strings.

File: src/sicks3000.cc (hunk 4 above)
Comment (ghost) [inline src/sicks3000.cc]: What's the benefit?  The reason for the macro is to reduce the possibility of a typo, and if one of the characters is wrong it needs only be fixed in one spot.

File: src/sicks3000.cc (hunk 3 above)
Comment (ghost) [inline src/sicks3000.cc]: I knew I was forgetting something!  Yeah, that was my intent actually, I think we talked about this too.  I'll toss this all into an array of structs now.

File: src/sicks3000.cc (hunk 5 above)
Comment (ghost) [inline src/sicks3000.cc]: They are the commands sent to the S3000, and they come from that document you linked in the JIRA.  I'll add some comments explaining this.

File: src/sicks3000.cc (hunk 6 above)
Comment (ghost) [inline src/sicks3000.cc]: Will update.

File: src/sicks3000.cc (hunk 7 above)
Comment (ghost) [inline src/sicks3000.cc]: Shouldn't happen, but there should be a case as you never know.  I'll add one.

Comment (ghost) [inline src/sicks3000.cc]: Anyway, the other important thing here is - where do we want to put this information?  I'm guessing we want this in the gui, so it should go out on a topic right?  Diagnostics?

File: src/sicks3000.cc [hunk 8]
```diff
@@ -310,3 +310,355 @@ unsigned short SickS3000::CreateCRC(uint8_t *Data, ssize_t length)
...
+{
+  const uint32_t error;
+  const char* cause;
+  const char* suggested_action;
+  
+} ErrorCodeLUTEntry;
+
+const ErrorCodeLUTEntry s3000_errors[1] = {
+  {SEVEN_SEGMENT_CODE(SEG_1_,SEG_0), "Object in protective field (1_.)", "No error. The status indication eases system testing on the use of simultaneous protective fields or in an EFI system (if the OSSDs on the guest are not used, a protective field infringement is not signaled via the red LED as required in the standard)."}
+};
```
Comment (ghost) [inline src/sicks3000.cc]: Before I go and convert everything, is this better?

Comment (mikepurvis) [inline src/sicks3000.cc]: Consider dropping the macro code and just having fields for `first_char`, `second_char`.

Drop the array length specifier— you don't need it.

File: src/sicks3000.cc [hunk 9]
```diff
@@ -310,3 +310,355 @@ unsigned short SickS3000::CreateCRC(uint8_t *Data, ssize_t length)
...
+      ROS_INFO("device_state %02X display_data %08X led_code %04X", device_state, display_data, (display_data & 0x0000FFFF));
+      
+      for(int i = 0; i < NUM_S3000_ERROR_CODES; i++)
+      {
+        if((display_data & 0x0000FFFF) == display_code[i])
+        {
+          ROS_INFO("%s", s3000_error_causes[i]);
+          ROS_INFO("%s", s3000_error_resolution[i]);
+          ros::Duration(0.1).sleep();
+          return 0;
```
Comment (ghost) [inline src/sicks3000.cc]: So what topic should this information be published on? Diagnostics?

Comment (mikepurvis) [inline src/sicks3000.cc]: See: https://github.com/clearpathrobotics/s3000_laser/blob/hydro-devel/src/s3000_laser.cc#L207

**Pull request closed without merging**